### DISCUSSION TITLE
Start an access point for initial configuration

### DISCUSSION THREAD

**Discussion by obones** [toverainc/willow#154] [Ideas]
Hello,

I believe this might already be "in the works" but I could not find it in the list "Future" section so I'm suggesting to have Willow create an access point the first time it runs or when booting while holding the top button pressed.
The display could show something like "Waiting for you on WILLOW wifi network" and a QR code to the captive portal URL so that the end user can then enter Wifi parameters and other options.
This would make for a friendlier user experience, especially when it comes to the final Wifi network and password that are currently hard coded in the firmware.
In personal ESP32 Arduino based projects, I have used  [IotWebConf](https://github.com/prampec/IotWebConf.git) which has its own limitations but is just a good illustration of what I have in mind:

![image](https://camo.githubusercontent.com/9c8fd28e4b81d607b27c65162900b49886acc97d598f4d91d363ad44be74fd6c/68747470733a2f2f736861726564696e76656e74696f6e732e636f6d2f77702d636f6e74656e742f75706c6f6164732f323031382f31312f53637265656e73686f745f32303138313130352d313931373438612e706e67)

I totally understand that because IotWebConf is Arduino based it is not usable in Willow but that could serve as an inspiration.

**Comment by kristiankielhofner** [comment]
The issue with this approach is that it's a little dicey for onboarding and it is very manual. Additionally, the typical use case of Willow is multiple devices and repeating the steps required with this approach is very tedious and error prone.

There's a good reason why (for example) ESPHome uses either a web flasher (via WebSerial) or requires direct serial connection for initial device configuration and flash.

It's still early but our initial thinking for this approach is a Tovera hosted WebSerial flasher that (other than initial loading of the web page) runs completely locally in the browser. Of course it could also be self-served by the user but there are issues related to TLS certs for WebSerial usage in browsers so self-hosting with this approach can be challenging. This is what ESPHome does today and I think it's a fantastic approach.

**Reply by obones** [reply]
Well, requiring a computer to do the initial setup is not something that I often see in consumer products, while an access point with a captive portal is quite common.

I do understand the requirement for ESPHome, in particular because it's targeted to a "maker"/"hacker" audience which I find acceptable.

But if I was an end user, with just a smartphone at hand, this kind of requirement would be a bit off putting.

**Reply by kristiankielhofner** [reply]
The issue is we don't ship a commercial product - Espressif does. Until they flash Willow at the factory connecting the device to a computer with a serial (USB) port will always be a requirement. While there, in your browser, you can enter your wifi details and keep clicking flash until you run out of devices. If something goes really wrong you can completely erase the device, re-flash, etc, etc with clicks in your browser.

From there WAS, dynamic configuration, and over the air updates handles the rest.

Even if Espressif were ever to flash Willow for us prior to shipment you still have the wifi configuration issue and I've always found the "Connect to an SSID, go to webpage, enter wifi details, reset and cross your fingers" approach to be pretty poor and error-prone.
